File: shared/fountain.ts
// wink 喷泉码（LT，Luby transform）—— 跨语言确定性实现
//
// 移植自 decimen-optical-transfer (MIT)。核心是跨语言一致性：
// - dlog：确定性自然对数，只用 IEEE-754 基础运算（禁止 Math.log）
// - splitmix32：纯整数运算（见 protocol.ts）
// - solitonCdf：robust-soliton 分布
// - frameIndices：逆 CDF 采样块子集
//
// Rust 端必须逐位复现（golden 向量断言锁定）。

import { splitmix32 } from './protocol';

const LN2 = 0.6931471805599453;
const SOLITON_C = 0.1;
const SOLITON_DELTA = 0.5;

/**
 * 确定性自然对数：精确 IEEE-754 运算（range reduction + atanh 级数）。
 * 不用 Math.log（libm 近似，JS 引擎之间/Rust 可能差 1 ulp，会静默失步）。
 */
export function dlog(x: number): number {
  let e = 0;
  let m = x;
  while (m >= 1.5) {
    m /= 2;
    e++;
  }
  while (m < 0.75) {
    m *= 2;
    e--;
  }
  const z = (m - 1) / (m + 1);
  const z2 = z * z;
  let term = z;
  let sum = 0;
  for (let n = 1; n <= 21; n += 2) {
    sum += term / n;
    term *= z2;
  }
  return e * LN2 + 2 * sum;
}

/** robust-soliton degree CDF for k source blocks */
export function solitonCdf(k: number): Float64Array {
  const cdf = new Float64Array(k);
  if (k === 1) {
    cdf[0] = 1;
    return cdf;
  }
  const R = Math.max(1, SOLITON_C * dlog(k / SOLITON_DELTA) * Math.sqrt(k));
  const spike = Math.min(k, Math.ceil(k / R));
  let total = 0;
  for (let d = 1; d <= k; d++) {
    const rho = d === 1 ? 1 / k : 1 / (d * (d - 1));
    let tau = 0;
    if (d < spike) tau = R / (d * k);
    else if (d === spike) tau = (R * Math.max(0, dlog(R / SOLITON_DELTA))) / k;
    total += rho + tau;
    cdf[d - 1] = total;
  }
  for (let i = 0; i < k; i++) cdf[i] = cdf[i]! / total;
  cdf[k - 1] = 1;
  return cdf;
}

/** 会话+序列号 → PRNG 种子 */
function frameSeed(sessionId: number, seq: number): number {
  let h = (Math.imul(sessionId + 1, 0x9e3779b1) ^ (seq + 0x85ebca6b)) | 0;
  h = Math.imul(h ^ (h >>> 13), 0xc2b2ae35);
  return (h ^ (h >>> 16)) | 0;
}

/**
 * 帧 seq 的块子集索引 —— 双端必须完全一致。
 * 逆 CDF 采样 degree + Fisher-Yates/去重采样块索引。
 */
export function frameIndices(
  k: number,
  cdf: Float64Array,
  sessionId: number,
  seq: number,
): number[] {
  const rnd = splitmix32(frameSeed(sessionId, seq));
  const u = rnd() * 2 ** -32;
  let lo = 0;
  let hi = k - 1;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (cdf[mid]! >= u) hi = mid;
    else lo = mid + 1;
  }
  const d = Math.min(k, lo + 1);
  if (d > k >> 3) {
    const scratch = new Uint32Array(k);
    for (let i = 0; i < k; i++) scratch[i] = i;
    const out: number[] = new Array<number>(d);
    for (let i = 0; i < d; i++) {
      const j = i + (rnd() % (k - i));
      const t = scratch[i]!;
      scratch[i] = scratch[j]!;
      scratch[j] = t;
      out[i] = scratch[i]!;
    }
    return out;
  }
  const set = new Set<number>();
  while (set.size < d) set.add(rnd() % k);
  return [...set];
}

function xorInto(dst: Uint32Array, src: Uint32Array): void {
  for (let i = 0; i < dst.length; i++) dst[i] = (dst[i]! ^ src[i]!) >>> 0;
}

export class LTEncoder {
  readonly k: number;
  private readonly words: number;
  private readonly blocks: Uint32Array;
  private readonly cdf: Float64Array;

  constructor(payload: Uint8Array, readonly blockLen: number, readonly sessionId: number) {
    this.k = Math.max(1, Math.ceil(payload.length / blockLen));
    this.words = Math.ceil(blockLen / 4);
    this.blocks = new Uint32Array(this.k * this.words);
    const bytes = new Uint8Array(this.blocks.buffer);
    for (let b = 0; b < this.k; b++) {
      const src = payload.subarray(b * blockLen, Math.min((b + 1) * blockLen, payload.length));
      bytes.set(src, b * this.words * 4);
    }
    this.cdf = solitonCdf(this.k);
  }

  encode(seq: number): Uint8Array {
    const idx = frameIndices(this.k, this.cdf, this.sessionId, seq);
    const out = new Uint32Array(this.words);
    for (const b of idx) {
      const off = b * this.words;
      for (let w = 0; w < this.words; w++) out[w] = (out[w]! ^ this.blocks[off + w]!) >>> 0;
    }
    return new Uint8Array(out.buffer, 0, this.blockLen);
  }

  /**
   * 找到生成 degree-1 帧且只包含目标块的 seq（重发指定块用）。
   * 扫描有限范围（默认 8192 个候选），找到即返回；找不到返回 null。
   * 该帧与普通帧完全同构，接收端用同一 seq 重新计算 indices → 直接解出该块。
   */
  findDeg1Seq(block: number, fromSeq = 0, scanLimit = 8192): number | null {
    if (block < 0 || block >= this.k) return null;
    for (let i = 0; i < scanLimit; i++) {
      const s = (fromSeq + i) >>> 0;
      const idx = frameIndices(this.k, this.cdf, this.sessionId, s);
      if (idx.length === 1 && idx[0] === block) return s;
    }
    return null;
  }

  /** 枚举所有 seq → degree-1 块映射（用于"已发送哪些块"视图） */
  deg1BlockAt(seq: number): number | null {
    const idx = frameIndices(this.k, this.cdf, this.sessionId, seq);
    if (idx.length === 1 && idx[0] !== undefined) return idx[0];
    return null;
  }
}

interface PendingFrame {
  idx: Set<number>;
  words: Uint32Array;
}

export class LTDecoder {
  private readonly words: number;
  private readonly cdf: Float64Array;
  private readonly solved: (Uint32Array | null)[];
  private readonly byBlock = new Map<number, Set<PendingFrame>>();
  private readonly seen = new Set<number>();
  solvedCount = 0;
  framesNew = 0;
  framesDup = 0;

  constructor(
    readonly k: number,
    readonly blockLen: number,
    readonly sessionId: number,
    readonly totalLen: number,
  ) {
    this.words = Math.ceil(blockLen / 4);
    this.cdf = solitonCdf(k);
    this.solved = new Array<Uint32Array | null>(k).fill(null);
  }

  get isComplete(): boolean {
    return this.solvedCount >= this.k;
  }

  /** 块状态：0=灰（未涉及）1=红（收到帧引用但未解出）2=绿（已解出） */
  blockState(b: number): 0 | 1 | 2 {
    if (this.solved[b]) return 2;
    if (this.byBlock.has(b)) return 1;
    return 0;
  }

  /** 已收到的帧 seq 位图（进度条按帧顺序显示用） */
  receivedSeqs(): Set<number> {
    return this.seen;
  }

  addFrame(seq: number, block: Uint8Array): void {
    if (this.seen.has(seq)) {
      this.framesDup++;
      return;
    }
    this.seen.add(seq);
    this.framesNew++;
    if (this.isComplete) return;

    const idx = new Set(frameIndices(this.k, this.cdf, this.sessionId, seq));
    const words = new Uint32Array(this.words);
    new Uint8Array(words.buffer).set(block.subarray(0, this.blockLen));
    for (const b of [...idx]) {
      const s = this.solved[b];
      if (s) {
        xorInto(words, s);
        idx.delete(b);
      }
    }
    if (idx.size === 0) return;
    if (idx.size === 1) {
      this.resolve(idx.values().next().value!, words);
      return;
    }
    const pf: PendingFrame = { idx, words };
    for (const b of idx) {
      let set = this.byBlock.get(b);
      if (!set) {
        set = new Set();
        this.byBlock.set(b, set);
      }
      set.add(pf);
    }
  }

  private resolve(b0: number, w0: Uint32Array): void {
    const queue: [number, Uint32Array][] = [[b0, w0]];
    while (queue.length > 0) {
      const [b, w] = queue.pop()!;
      if (this.solved[b]) continue;
      this.solved[b] = w;
      this.solvedCount++;
      const waiting = this.byBlock.get(b);
      if (!waiting) continue;
      this.byBlock.delete(b);
      for (const pf of waiting) {
        xorInto(pf.words, w);
        pf.idx.delete(b);
        if (pf.idx.size === 1) {
          const r = pf.idx.values().next().value!;
          this.byBlock.get(r)?.delete(pf);
          if (!this.solved[r]) queue.push([r, pf.words]);
        }
      }
    }
  }

  assemble(): Uint8Array | null {
    if (!this.isComplete) return null;
    const out = new Uint8Array(this.totalLen);
    for (let b = 0; b < this.k; b++) {
      const start = b * this.blockLen;
      const len = Math.min(this.blockLen, this.totalLen - start);
      if (len > 0) out.set(new Uint8Array(this.solved[b]!.buffer, 0, len), start);
    }
    return out;
  }
}

File: sender-cli/src/fountain.rs
// wink 喷泉码（Rust 端）—— 与 shared/fountain.ts 逐位一致
//
// 跨语言铁律：
// - dlog: 只用 IEEE-754 基础运算（+ - * /），禁止 f64::ln
// - soliton_cdf: 同参数 C=0.1 δ=0.5
// - frame_indices: 逆 CDF 二分 + Fisher-Yates/去重采样
//
// clippy 说明：协议路径的 `as f64` cast 是安全的（k ≤ 65535、i32 范围），
// 为保持与 TS 移植的可读性，局部 allow 而非全改 From。

#![allow(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss,
    clippy::cast_possible_wrap,
    clippy::many_single_char_names,
    clippy::cast_lossless,
    clippy::too_many_lines,
    clippy::module_name_repetitions
)]

use crate::protocol::SplitMix32;

const LN2: f64 = std::f64::consts::LN_2;
const SOLITON_C: f64 = 0.1;
const SOLITON_DELTA: f64 = 0.5;

/// 确定性自然对数（与 TS dlog 逐位一致）
#[must_use]
pub fn dlog(x: f64) -> f64 {
    let mut e: i32 = 0;
    let mut m = x;
    while m >= 1.5 {
        m /= 2.0;
        e += 1;
    }
    while m < 0.75 {
        m *= 2.0;
        e -= 1;
    }
    let z = (m - 1.0) / (m + 1.0);
    let z2 = z * z;
    let mut term = z;
    let mut sum = 0.0;
    let mut n = 1;
    while n <= 21 {
        sum += term / n as f64;
        term *= z2;
        n += 2;
    }
    (e as f64) * LN2 + 2.0 * sum
}

/// robust-soliton degree CDF
#[must_use]
pub fn soliton_cdf(k: usize) -> Vec<f64> {
    let mut cdf = vec![0.0f64; k];
    if k == 1 {
        cdf[0] = 1.0;
        return cdf;
    }
    let r = (SOLITON_C * dlog(k as f64 / SOLITON_DELTA) * (k as f64).sqrt()).max(1.0);
    let spike = ((k as f64) / r).ceil().min(k as f64) as usize;
    let mut total = 0.0;
    for d in 1..=k {
        let rho = if d == 1 {
            1.0 / k as f64
        } else {
            1.0 / ((d * (d - 1)) as f64)
        };
        let mut tau = 0.0;
        if d < spike {
            tau = r / (d as f64 * k as f64);
        } else if d == spike {
            tau = (r * (dlog(r / SOLITON_DELTA)).max(0.0)) / k as f64;
        }
        total += rho + tau;
        cdf[d - 1] = total;
    }
    for v in &mut cdf {
        *v /= total;
    }
    cdf[k - 1] = 1.0;
    cdf
}

fn frame_seed(session_id: u16, seq: u32) -> u32 {
    // TS: h = imul(sessionId+1, 0x9e3779b1) ^ (seq + 0x85ebca6b)
    let h = (session_id as u32)
        .wrapping_add(1)
        .wrapping_mul(0x9e37_79b1)
        ^ seq.wrapping_add(0x85eb_ca6b);
    // TS: h = imul(h ^ (h>>>13), 0xc2b2ae35)
    let h = (h ^ (h >> 13)).wrapping_mul(0xc2b2_ae35);
    // TS: return (h ^ (h>>>16)) | 0
    h ^ (h >> 16)
}

/// 帧 seq 的块子集（与 TS frameIndices 逐位一致）
#[must_use]
pub fn frame_indices(k: usize, cdf: &[f64], session_id: u16, seq: u32) -> Vec<usize> {
    let mut rnd = SplitMix32::new(frame_seed(session_id, seq));
    let u = rnd.next_u32() as f64 * 2f64.powi(-32);
    let mut lo = 0usize;
    let mut hi = k - 1;
    while lo < hi {
        let mid = usize::midpoint(lo, hi);
        if cdf[mid] >= u {
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    let d = (lo + 1).min(k);
    if d > k >> 3 {
        // 大 degree：部分 Fisher-Yates
        let mut scratch: Vec<usize> = (0..k).collect();
        let mut out = Vec::with_capacity(d);
        for i in 0..d {
            let j = i + (rnd.next_u32() as usize % (k - i));
            scratch.swap(i, j);
            out.push(scratch[i]);
        }
        out
    } else {
        // 小 degree：去重采样
        let mut set = std::collections::HashSet::new();
        while set.len() < d {
            set.insert(rnd.next_u32() as usize % k);
        }
        set.into_iter().collect()
    }
}

pub struct LTEncoder {
    pub k: usize,
    words: usize,
    blocks: Vec<u32>,
    cdf: Vec<f64>,
    pub block_len: usize,
    session_id: u16,
}

impl LTEncoder {
    #[must_use]
    pub fn new(payload: &[u8], block_len: usize, session_id: u16) -> Self {
        let k = payload.len().div_ceil(block_len).max(1);
        let words = block_len.div_ceil(4);
        let mut blocks = vec![0u32; k * words];
        for (b, chunk) in payload.chunks(block_len).enumerate() {
            let base = b * words * 4;
            for (i, byte) in chunk.iter().enumerate() {
                blocks[(base + i) / 4] |= (*byte as u32) << ((i % 4) * 8);
            }
        }
        let cdf = soliton_cdf(k);
        Self {
            k,
            words,
            blocks,
            cdf,
            block_len,
            session_id,
        }
    }

    #[must_use]
    pub fn encode(&self, seq: u32) -> Vec<u8> {
        let idx = frame_indices(self.k, &self.cdf, self.session_id, seq);
        let mut out = vec![0u32; self.words];
        for &b in &idx {
            let off = b * self.words;
            for (w, item) in out.iter_mut().enumerate() {
                *item ^= self.blocks[off + w];
            }
        }
        // block_len 可能不是 4 的倍数（自动优化算出 370 等）：
        // words 是 ceil(block_len/4)，写 bytes 时截断到 block_len，防越界
        let mut bytes = vec![0u8; self.block_len];
        for (i, w) in out.iter().enumerate() {
            let base = i * 4;
            for j in 0..4 {
                let pos = base + j;
                if pos >= self.block_len {
                    break;
                }
                bytes[pos] = (*w >> (j * 8)) as u8;
            }
        }
        bytes
    }

    /// 找到生成 degree-1 帧且只包含目标块的 seq（重发指定块用）
    /// 扫描有限范围（默认 8192 个候选），找到即返回；找不到返回 None
    #[must_use]
    pub fn find_deg1_seq(&self, block: usize, from_seq: u32, scan_limit: usize) -> Option<u32> {
        if block >= self.k {
            return None;
        }
        for i in 0..scan_limit {
            let s = from_seq.wrapping_add(i as u32);
            let idx = frame_indices(self.k, &self.cdf, self.session_id, s);
            if idx.len() == 1 && idx[0] == block {
                return Some(s);
            }
        }
        None
    }

    /// 找到包含目标块的任意帧 seq（degree-1 稀疏时的降级方案）
    #[must_use]
    pub fn find_any_seq(&self, block: usize, from_seq: u32, scan_limit: usize) -> Option<u32> {
        if block >= self.k {
            return None;
        }
        for i in 0..scan_limit {
            let s = from_seq.wrapping_add(i as u32);
            let idx = frame_indices(self.k, &self.cdf, self.session_id, s);
            if idx.contains(&block) {
                return Some(s);
            }
        }
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dlog_known_values() {
        assert_eq!(dlog(1.0), 0.0);
        assert_eq!(dlog(2.0), 0.6931471805599453);
        assert_eq!(dlog(std::f64::consts::E), 1.0);
        for x in [0.5f64, 1.5, 10.0, 100.0, 0.25] {
            assert!((dlog(x) - x.ln()).abs() < 1e-12, "dlog({x})");
        }
    }

    #[test]
    fn soliton_k1() {
        assert_eq!(soliton_cdf(1), vec![1.0]);
    }

    #[test]
    fn soliton_k100_sane() {
        let cdf = soliton_cdf(100);
        assert_eq!(cdf.len(), 100);
        assert_eq!(cdf[99], 1.0);
        for i in 1..100 {
            assert!(cdf[i] >= cdf[i - 1], "cdf[{i}] monotone");
        }
    }

    #[test]
    fn frame_indices_deterministic_in_range() {
        let k = 8;
        let cdf = soliton_cdf(k);
        let a = frame_indices(k, &cdf, 1, 0);
        let b = frame_indices(k, &cdf, 1, 0);
        assert_eq!(a, b);
        for idx in &a {
            assert!(*idx < k);
        }
    }

    #[test]
    fn lt_frames_deterministic_and_distinct() {
        let payload: Vec<u8> = (0..2048).map(|i| (i * 7 % 256) as u8).collect();
        let block_len = 64;
        let enc = LTEncoder::new(&payload, block_len, 7);
        for seq in 0..enc.k * 2 {
            let frame = enc.encode(seq as u32);
            assert_eq!(frame.len(), block_len);
        }
        assert_ne!(enc.encode(0), enc.encode(1));
    }

    #[test]
    fn encode_non_multiple_of_4_block_len_no_panic() {
        // 回归：block_len=370（非 4 倍数，自动优化可能算出）曾导致越界 panic
        let payload: Vec<u8> = (0..5000).map(|i| (i * 13 % 256) as u8).collect();
        let block_len = 370;
        let enc = LTEncoder::new(&payload, block_len, 42);
        for seq in 0..enc.k * 2 {
            let frame = enc.encode(seq as u32);
            assert_eq!(frame.len(), block_len, "frame len = block_len");
        }
        assert_ne!(enc.encode(0), enc.encode(1));
        // 确定性
        assert_eq!(enc.encode(7), enc.encode(7));
    }
}
